### ISSUE TITLE
IS-07: Add WebSocket receiver

### ISSUE THREAD

**Issue by garethsb-sony** [sony/nmos-cpp#73]
PR #69 added support for IS-07 Event & Tally senders using the WebSocket transport.

It would be nice to add support for receivers, using the cpprestsdk `web::websockets::client::ws_client`. We've already used this elsewhere for a Query WebSocket client.

The tricky bit may be deciding how to store the received data... perhaps it needs a new `nmos::types::destination` type to be used in the `event_resources` container of the `nmos::node_model`, which would have `"state"` mirroring the source? Notifications of updates would be available via the usual model `condition` variable.

**Comment by jwpwh** [comment]
Is the code using the Query WebSocket client contained within this nmos-cpp project? If not, is it available elsewhere?

**Comment by garethsb-sony** [comment]
Sorry, no, I was referring to some code I can't easily extract from where it is; are you more interested in Event & Tally WebSocket Receiver or Query WebSocket client?

**Comment by jwpwh** [comment]
I am interested in both, in the long term. Right now I am looking more into the Query WebSocket client. I realize my interest is probably outside the scope of the sample node code which is offered in this project.

I am working on making an existing system be able to offer audio to a system using NMOS protocols. The sample node here has been great for this. The Event & Tally WebSocket Receiver will be needed for this part of the project.

I am also working on making our existing system be the control and admin component of a system using our proprietary devices as well as AES67 devices that might use the NMOS protocols.  For this aspect, the Query WebSocket client, and more, is needed so we can discover all the devices and make connections. Probably need to use technology similar to the Riedel NMOS explorer.

I thought it would be worth asking just in case something was available.

Thanks.

**Comment by garethsb-sony** [comment]
Thanks @jwpwh, I appreciate your sharing that. I will try to carve out some time to put a basic C++ Query WebSocket client together if I can't extract the code that exists 'elsewhere'!

**Comment by jwpwh** [comment]
Thanks. A basic client would be great. If I can help make it more complete, I would be glad to do that.

**Comment by garethsb-sony** [comment]
What do you think would make a reasonable minimal example app?

Simplest possible would just make a subscription, make a WebSocket connection and log add/remove/change/sync events to console so long as the connection was open, I suppose.

More interesting might be to maintain a local `nmos::resources` container based on the add/remove/change/sync events.

Other ideas?

**Comment by jwpwh** [comment]
I can see needing all of what you have listed. In addition eventually working in peer-to-peer and getting connection information from the nodes themselves. The first step you mention with making a subscription and logging to the console would be a good start.

**Comment by garethsb-sony** [comment]
OK, thanks. Consider it on the to-do list.

> In addition eventually working in peer-to-peer ...

Nodes don't support a WebSocket API, so the population of a local 'Query API' in peer-to-peer operation (as suggested by the spec) would have to use the mDNS `_nmos-node._tcp` advertisements and then the Node API, with updates triggered by changes to the `ver_` TXT records or Node advertisements coming and going. The target could be `nmos::resources`, but apart from that there wouldn't be much implementation overlap with the Query WebSocket API usage, unfortunately.

**Comment by garethsb-sony** [comment]
Coincidentally, @bakaleks and I have been discussing support for WebSocket ping-pong. If we want to show this off in a client, we'll need https://github.com/microsoft/cpprestsdk/pull/1143.

**Comment by garethsb-sony** [comment]
> Coincidentally, @bakaleks and I have been discussing support for WebSocket ping-pong. If we want to show this off in a client, we'll need https://github.com/microsoft/cpprestsdk/pull/1143.

Now merged to cpprestsdk master.

**Comment by garethsb-sony** [comment]
@bakaleks made a good point to me at the AMWA workshop that an implementation may want to guarantee that it processes all events. This means that a simple `destination` type holding a _single_ most recent `"state"` proposed in original comment isn't enough for all purposes. The design needs to allow off-loading of opening the WebSocket connection, with an example using `ws_client` to do something like populate a local `source` (whose `"parents"` could be set appropriately).

**Comment by garethsb-sony** [comment]
C++ REST SDK v2.10.14 is tagged. I haven't tested the release yet.

I'm working on nmos module code to manage WebSocket connections for multiple Receivers subscribed to the same or different `connection_uri` as per the IS-07 spec, and an example Receiver to match the existing example Sender.

**Comment by garethsb-sony** [comment]
I pushed the WebSocket receiver implementation yesterday (27c6a0ed839b5900d99cc7b34c701baa686e3e47) and have tested it with a couple of Senders and a couple of Receivers each on a couple of Nodes, to prove the sharing (or not) of WebSocket connections depending on the `"connection_uri"`.

The only missing feature is the under-specified "behind the scenes" reconnection logic (see second paragraph of [Message types - The health message type](https://github.com/AMWA-TV/nmos-event-tally/blob/v1.0.x/docs/2.0.%20Message%20types.md#15-the-health-message-type) and third paragraph of [WebSocket Transport - Subscriptions strategy](https://github.com/AMWA-TV/nmos-event-tally/blob/v1.0.x/docs/5.2.%20Transport%20-%20Websocket.md#4-subscriptions-strategy)).

I'm closing this issue, but I've opened another to cover an example Query WebSocket API client.
